containers/hawkular-openshift-security-filter/src/main/java/org/hawkular/openshift/auth/org/hawkular/openshift/namespace/NamespaceHandler.java

```diff
@@ -0,0 +1,128 @@
+import org.jboss.logging.Logger;
+
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.HttpString;
+
+/**
+ *
+ */
+public class NamespaceHandler implements HttpHandler {
+
+    private static final Logger log = Logger.getLogger(NamespaceHandler.class);
+
+    private final HttpHandler containerHandler;
+
+    //
+    private static final String HAWKULAR_TENANT_HEADER_NAME = "Hawkular-Tenant";
+    private static final String TOKEN_FILE_LOCATION = "/var/run/secrets/kubernetes.io/serviceaccount/token";
+
+    private static final String KUBERNETES_MASTER_URL_SYSPROP = "KUBERNETES_MASTER_URL";
+    private static final String KUBERNETES_MASTER_URL_DEFAULT = "https://kubernetes.default.svc.cluster.local";
+    static final String KUBERNETES_MASTER_URL =
+            System.getProperty(KUBERNETES_MASTER_URL_SYSPROP, KUBERNETES_MASTER_URL_DEFAULT);
+    //
+
+    private final String token;
+
+    private final NamespaceListener namespaceListener;
+
+    private final Map<String,String> projectOverwrites;
+
+    public NamespaceHandler(HttpHandler containerHandler) {
+        log.info("Creating NamespaceHandler");
+        this.containerHandler = containerHandler;
+
+        try {
+            this.token = getToken();
+            log.info("Received token from " + TOKEN_FILE_LOCATION);
+        } catch (Exception e){
+            log.info("Error retrieving token from file " + TOKEN_FILE_LOCATION);
+            throw new RuntimeException(e);
+        }
+
+        namespaceListener = new NamespaceListener(token);
+        NamespaceOverrideMapper nom = new NamespaceOverrideMapper(token);
+        log.info("Received namespace overrides: " + nom.overrides);
+
+        this.projectOverwrites = nom.overrides;
+    }
+
+    private String getToken() throws IOException {
+        return new String(Files.readAllBytes(Paths.get(TOKEN_FILE_LOCATION)));
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange serverExchange) throws Exception {
+        String originalTenantName = getTenantName(serverExchange);
+        if (originalTenantName == null || originalTenantName.startsWith("_")) {
+            containerHandler.handleRequest(serverExchange);
+        } else {
+
+            String namespaceName = "";
+            String[] splitted = originalTenantName.split(",");
+            for (int i=0; i<splitted.length; i++) {
+                String namespaceID = namespaceListener.getNamespaceId(splitted[i]);
+                if (namespaceID != null) {
+                    String tenantName;
+
+                    if (projectOverwrites.containsKey(namespaceID)) {
+                        tenantName = projectOverwrites.get(namespaceID);
+                    } else {
+                        tenantName = splitted[i] + ":" + namespaceID;
+                    }
+
+                    if (i < splitted.length - 1) {
+                        tenantName += ",";
```

Comment:
It works fine for me with JDK 8. I will update it to use StringJoiner as it does simplify things.